### PR TITLE
DIDman: Strict API interfaces

### PR DESCRIPTION
**Pull request by reinkrul**
Still causes compilation issues, which I fixed in the codegen in https://github.com/deepmap/oapi-codegen/pull/999, but which needs to be released. For the time being, maybe just fix it manually (it's changing 1 character), we don't alter the API too often.

### CODE REVIEW

**Comment by gerardsn** [comment]
title should be `Didman` instead of `VDR`?

**Comment by reinkrul** [comment]
Things left to do:
- [ ] Find out how to get request Content-Type header (failing test)
- [ ] Test `Routes`

File: didman/api/v1/api.go [hunk 1]
```diff
@@ -77,81 +77,88 @@ func (w *Wrapper) ResolveStatusCode(err error) int {
...
+	id, err := did.ParseDID(request.Did)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	if len(strings.TrimSpace(request.Type)) == 0 {
-		return core.InvalidInputError("invalid value for type")
+	props := request.Body
+	if len(strings.TrimSpace(props.Type)) == 0 {
```
Comment (gerardsn) [inline didman/api/v1/api.go:106]: I prefer the explicitness of `request.Body.Type`

File: didman/api/v1/api.go [hunk 2]
```diff
@@ -162,72 +169,55 @@ func (w *Wrapper) GetCompoundServices(ctx echo.Context, didStr string) error {
...
```
Comment (gerardsn) [inline didman/api/v1/api.go:179]: same here

File: didman/api/v1/api_test.go [hunk 3]
```diff
@@ -34,462 +32,395 @@ import (
...
 
 	t.Run("ok", func(t *testing.T) {
-		ctx := newMockContext(t)
-		ctx.didman.EXPECT().DeleteEndpointsByType(audit.ContextWithAuditInfo(), *parsedID, endpointType)
-		ctx.echo.EXPECT().NoContent(http.StatusNoContent)
+		test := newMockContext(t)
+		test.didman.EXPECT().DeleteEndpointsByType(audit.ContextWithAuditInfo(), *parsedID, endpointType)
 
-		err := ctx.wrapper.DeleteEndpointsByType(ctx.echo, idStr, endpointType)
+		_, err := test.wrapper.DeleteEndpointsByType(ctx, request)
```
Comment (gerardsn) [inline didman/api/v1/api_test.go:188]: check that response is not nil

File: didman/api/v1/api.go [hunk 4]
```diff
@@ -162,72 +169,55 @@ func (w *Wrapper) GetCompoundServices(ctx echo.Context, didStr string) error {
...
+		return nil, err
 	}
 
-	// default json
-	return ctx.JSON(http.StatusOK, EndpointResponse{Endpoint: endpoint})
+	if request.Params.Accept == nil || *request.Params.Accept != "text/plain" {
+		return GetCompoundServiceEndpoint200JSONResponse{Endpoint: endpoint}, nil
+	} else {
+		return GetCompoundServiceEndpoint200TextResponse(endpoint), nil
+	}
```
Comment (gerardsn) [inline didman/api/v1/api.go:220]: explicitly test for `text/plain` or default to json outside the if statement reduces cognitive complexity

File: didman/api/v1/api_test.go [hunk 5]
```diff
@@ -34,462 +32,395 @@ import (
...
-		assert.EqualError(t, err, "failed to parse v1.CompoundServiceProperties: b00m!")
+		assert.Equal(t, http.StatusBadRequest, test.wrapper.ResolveStatusCode(err))
+		assert.Nil(t, response)
 	})
 }
 
 func TestWrapper_GetCompoundServices(t *testing.T) {
-	idStr := "did:nuts:1#1"
-	id, _ := did.ParseDIDURL(idStr)
+	targetDID := did.MustParseDID("did:nuts:1")
```
Comment (gerardsn) [inline didman/api/v1/api_test.go:327]: test has changed by removing the fragment

File: docs/_static/didman/v1.yaml
Comment (gerardsn) [inline docs/_static/didman/v1.yaml:1]: nice find! I was looking for a solution for this

**Review by gerardsn — CHANGES_REQUESTED** [review]
just some minor things

File: didman/api/v1/api_test.go (hunk 5 above)
Comment (gerardsn) [reply didman/api/v1/api_test.go:327]: the GetCompoundServices endpoint parses it as `id, err := did.ParseDIDURL(request.Did)` though. Is that wrong too then?

Comment (reinkrul) [reply didman/api/v1/api_test.go:327]: fixed

**Review by gerardsn — APPROVED** [review]

**Pull request merged**